pkg/tracing/context.go

```diff
@@ -0,0 +1,41 @@
+package tracing
+
+import (
+	"context"
+
+	"github.com/opentracing/opentracing-go"
+)
+
+type contextKey struct{}
+
+var tracerKey = contextKey{}
+
+// ContextWithSpan returns a new `context.Context` that holds a reference to given opentracing.Tracer.
+func ContextWithTracer(ctx context.Context, tracer opentracing.Tracer) context.Context {
+	return context.WithValue(ctx, tracerKey, tracer)
+}
+
+func tracerFromContext(ctx context.Context) opentracing.Tracer {
+	val := ctx.Value(tracerKey)
+	if sp, ok := val.(opentracing.Tracer); ok {
+		return sp
+	}
+	return nil
+}
+
+// StartSpanFromContext starts and returns span with `operationName` using any Span found within given context.
+// It uses traces propagated in context.
+func StartSpanFromContext(ctx context.Context, operationName string, opts ...opentracing.StartSpanOption) (opentracing.Span, context.Context) {
+	tracer := tracerFromContext(ctx)
+	if tracer == nil {
+		// No tracing found, use noop one.
+		tracer = &opentracing.NoopTracer{}
+	}
+
+	var span opentracing.Span
+	if parentSpan := opentracing.SpanFromContext(ctx); parentSpan != nil {
+		opts = append(opts, opentracing.ChildOf(parentSpan.Context()))
+	}
+	span = tracer.StartSpan(operationName, opts...)
+	return span, opentracing.ContextWithSpan(ctx, span)
```

Comment:
I see that we are often ignoring the returned context when calling this function. My intuition for this API would be that I'd always overwrite my calling context with it like `span, ctx := StartSpanFromContext(ctx, ...)`. Does this not make sense? How would it make sense to use this context?

Reply:
Yes, you are right. And actually in 80% places we are doing that. The only exception is function like:
```
{
span, _ := StartSpanFromContext(ctx, ...)
// some A stuff...
span.Finish()

span, _ := StartSpanFromContext(ctx, ...)
// some B stuff... 
span.Finish()
```
We don't overwrite context, because we don't want to make stuff A a parent of stuff B, right? (: